dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -175,10 +178,82 @@ public Version parseWorkflowContent(String filePath, String content, Set<SourceF
                 validationMessageObject.put(filePath, "CWL file is malformed or missing, cannot extract metadata: " + message);
                 version.addOrUpdateValidation(new Validation(DescriptorLanguage.FileType.DOCKSTORE_CWL, false, validationMessageObject));
             }
+
         }
         return version;
     }
 
+    /**
+     * Examine the specified set of SourceFiles for CWL language versions,
+     * set the language version of each SourceFile, and set the list of
+     * language versions of the specified Version.
+     *
+     * In CWL, each SourceFile can contain multiple tools or subworkflows,
+     * each with its own version.  If a SourceFile specifies multiple CWL
+     * versions, the language version is set to the "latest" version per
+     * the list at the bottom of:
+     * https://www.commonwl.org/v1.2/CommandLineTool.html#CWLVersion
+     *
+     * The list of language versions for a given Version is sorted
+     * in descending order, from "latest" to "earliest".
+     */
+    private void setCwlVersionsFromSourceFiles(Set<SourceFile> sourceFiles, Version version) {
+        Set<String> allVersions = new HashSet<>();
+        for (SourceFile file: sourceFiles) {
+            if (file.getType() == DescriptorLanguage.FileType.DOCKSTORE_CWL) {
+                Set<String> fileVersions = getCwlVersionsFromSourceFile(file);
+                file.setTypeVersion(newestVersion(fileVersions));
+                allVersions.addAll(fileVersions);
+            }
+        }
+        version.setDescriptorTypeVersions(sortVersionsDescending(allVersions));
+    }
+
+    /**
+     * Extract the set of CWL language versions that appear in the
+     * specified SourceFile.
+     */
+    private Set<String> getCwlVersionsFromSourceFile(SourceFile file) {
+        Set<String> versions = new HashSet<>();
+        getCwlVersionsFromMap(versions, parseAsMap(file.getContent()));
+        return versions;
+    }
+
+    /**
+     * Recursively examine the parsed YAML Map/List representation of
+     * a CWL file, interpreting the string value of any Map entry with
+     * the key "cwlVersion" as a language version.
+     */
+    private void getCwlVersionsFromMap(Set<String> versions, Map<String, Object> map) {
+        map.forEach((k, v) -> {
+            if (Objects.equals(k, "cwlVersion") && v instanceof String) {
+                versions.add(v.toString());
```

Comment:
It looks like we don't do any validation of the version, which I suppose is OK. E.g., if you had `cwlVersion: "I am a bad example of a version"`, `I am a bad example of a version` would be the version?

I'm asking because I'm wondering how to handle invalid versions in my upcoming similar Nextflow ticket.

Reply:
> It looks like we don't do any validation of the version, which I suppose is OK. E.g., if you had `cwlVersion: "I am a bad example of a version"`, `I am a bad example of a version` would be the version?
> 
> I'm asking because I'm wondering how to handle invalid versions in my upcoming similar Nextflow ticket.

Yeah, a possible solution, for the purposes of calculating the language version[s], might be to simply ignore the "weird" (garbagey, unknown, non-standard) versions.  In the case of CWL, maybe "weird" could be defined as anything that's not of the form 'vD.D' (where D represents one or more digits).  Perhaps, we'd also consider also suffixed versions of the former as non-weird (which we could map to the non-suffixed `vD.D` value if we though it was appropriate)?

FWIW, here's the list of `cwlVersion` values from sourcefiles in our database (as of a few months ago):
```
cwl:draft-3
cwl:draft-3.dev3
cwl:v1.0
draft-3
latest
nonvalid
sbg:draft-2
v1.0
v1.1
v1.2
```

Reply:
> I'm asking because I'm wondering how to handle invalid versions in my upcoming similar Nextflow ticket.

For WDL, I only recorded valid versions: https://github.com/dockstore/dockstore/blob/4a0895370d4e62f1dfeccb86b7f04ed35f064507/dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java#L586

My thought was that if a descriptor sourcefile or entry version doesn't have language versions set, it likely means that the language version is invalid (weird values or missing)

Reply:
> For WDL, I only recorded valid versions:  

Interesting, we might have to revisit that if/we when add miniwdl support, e.g., what if if miniwdl supports 1.1 but Cromwell does not? All hypothetical, and no action required, just noting it for percolation purposes.